xwayland/selection/incoming.c

```diff
@@ -150,6 +150,21 @@ static void source_send(struct wlr_xwm_selection *selection,
 		return;
 	}
 
+	// FIXME: we currently can't handle two X11-to-Wayland transfers at once due
+	// to reusing the same X11 window. Proceeding further here would lead us to
+	// lose track of the current `transfer->wl_client_fd` and use-after-free
+	// during cleanup. This doesn't happen often, but bail now to avoid a
+	// compositor crash later.
+	if (transfer->wl_client_fd > 0) {
```

Comment:
Zero is a valid value for an FD. If the compositor closes stdin, FD 0 could be re-used for a selection.

Reply:
You're right, but 0 is also what `wl_client_fd` is "set" to post-`calloc`, so including 0 in this check as-is would totally break copy/paste (the very first transfer would perpetually be rejected). I've introduced some more logic to initialize `wl_client_fd` to -1. In the future, we can use `xwm_selection_transfer_init` to do the X11 window allocation stuff, too.

Reply:
Ah, good catch!